qemu/tests/invalid_cpu_device_hotplug.py

```diff
@@ -0,0 +1,129 @@
+
+    :param test: QEMU test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment.
+    """
+    not_match_err = ("Hotplug %s failed but the error description does not "
+                     "match: '%s'")
+    expected_info = "Hotplug %s failed as expected, error description: '%s'"
+    hotplug_pass_err = "Still able to hotplug %s via qmp"
+
+    def hotplug_inuse_vcpu():
+        """Hotplug 2 in use vcpu devices: main vcpu and duplicate vcpu"""
+        # Main vCPU
+        main_vcpu_props = {}
+        for vcpu_prop in vcpu_props:
+            main_vcpu_props.setdefault(vcpu_prop, "0")
+        vm.params["vcpu_props_main_vcpu"] = json.dumps(main_vcpu_props)
+        error_context.context("Define the invalid vcpu: %s" % "main_vcpu",
+                              logging.info)
+        in_use_vcpu_dev = vm.devices.vcpu_device_define_by_params(vm.params,
+                                                                  "main_vcpu")
+        try:
+            error_context.context("Hotplug the main vcpu", logging.info)
+            in_use_vcpu_dev.enable(vm.monitor)
+        except QMPCmdError as err:
+            qmp_desc = err.data["desc"]
+            if error_desc.format("0") not in qmp_desc:
+                test.error(not_match_err % ("main vcpu", qmp_desc))
+            logging.info(expected_info, "main vcpu", qmp_desc)
+        else:
+            test.fail(hotplug_pass_err % "main vcpu")
+
+        # New vCPU
+        error_context.context("hotplug vcpu device: %s" % vcpu_device_id,
+                              logging.info)
+        vm.hotplug_vcpu_device(vcpu_device_id)
+        if vm.get_cpu_count() != maxcpus:
+            test.fail("Actual number of guest CPUs is not equal to expected.")
+
+        # Duplicate vCPU
+        duplicate_vcpu_params = vm.devices.get_by_qid(
+            vcpu_device_id)[0].params.copy()
+        del duplicate_vcpu_params["id"]
+        vm.params["vcpu_props_duplicate_vcpu"] = json.dumps(
+            duplicate_vcpu_params)
+        duplicate_vcpu_dev = vm.devices.vcpu_device_define_by_params(
+            vm.params, "duplicate_vcpu")
+        try:
+            error_context.context("hotplug the duplicate vcpu", logging.info)
+            duplicate_vcpu_dev.enable(vm.monitor)
+        except QMPCmdError as err:
+            dev_count = maxcpus
+            if 'ppc64' in arch_name:
+                dev_count //= threads
+            qmp_desc = err.data["desc"]
+            if error_desc.format(str(dev_count - 1)) not in qmp_desc:
+                test.error(not_match_err % ("duplicate vcpu", qmp_desc))
+            logging.info(expected_info, "duplicate vcpu", qmp_desc)
+        else:
+            test.fail(hotplug_pass_err % "duplicate vcpu")
+
+    def hotplug_invalid_vcpu():
+        """Hotplug a vcpu device with invalid property id"""
+        vcpu_device = vm.devices.get_by_qid(vcpu_device_id)[0]
+        vm.devices.remove(vcpu_device)
```

Comment:
Why do we need remove the device?

Reply:
We define a valid vcpu device in all cases to initialize the vcpu bus, for duplicate vcpu, we can hotplug it, for other cases, we need to remove it and change it's params since we want to hotplug an invalid device.

Reply:
Makes sense to me, thanks.